### ISSUE TITLE
Service continues and fails to configure an ELB, locking out AWS account

### ISSUE THREAD

**Issue by evanphx** [kubernetes/kubernetes#11324]
After setting up a few services, I noticed the following output in `get events`:

```
Tue, 14 Jul 2015 18:04:43 -0700   Wed, 15 Jul 2015 11:26:22 -0700   3544      my-nginx                                     Service               creating loadbalancer failed   {service-controller }                                  failed to create external load balancer for service default/my-nginx: error listing AWS subnets: RequestLimitExceeded: Request limit exceeded.
                                  status code: 503, request id: []
Tue, 14 Jul 2015 18:04:44 -0700   Wed, 15 Jul 2015 11:26:22 -0700   2679      my-nginx   Service             creating loadbalancer failed   {service-controller }   failed to create external load balancer for service default/my-nginx: error listing AWS security groups: RequestLimitExceeded: Request limit exceeded.
                                  status code: 503, request id: []
Tue, 14 Jul 2015 18:04:08 -0700   Wed, 15 Jul 2015 11:26:31 -0700   4926      my-nginx   Service             creating loadbalancer failed   {service-controller }   failed to create external load balancer for service default/my-nginx: error listing AWS VPCs: RequestLimitExceeded: Request limit exceeded.
                                  status code: 503, request id: []
Tue, 14 Jul 2015 18:01:05 -0700   Wed, 15 Jul 2015 11:26:39 -0700   67946     my-nginx   Service             creating loadbalancer failed   {service-controller }   failed to create external load balancer for service default/my-nginx: error listing AWS instances: RequestLimitExceeded: Request limit exceeded.
                                  status code: 503, request id: []
Tue, 14 Jul 2015 15:55:48 -0700   Wed, 15 Jul 2015 11:26:42 -0700   25049     my-nginx   Service             creating loadbalancer failed   {service-controller }   failed to create external load balancer for service default/my-nginx: InvalidGroup.Duplicate: The security group 'k8s-elb-a744c2a772a7b11e593680276d36fbdd' already exists for VPC 'vpc-f661fa93'
                                  status code: 400, request id: []
```

I assume the service tried to configure an ELB in the past (not sure when or how or what details it tried to use) and has kept trying and now can't even use the AWS API.

How can I either:
1) tell it to please stop so that I can use my AWS account again
2) fix it so that it does configure the ELB

/cc @justinsb

**Comment by justinsb** [comment]
Sorry about the problem.  Looks like we need a sleep / backoff policy here!

To get you up and running again, if you redeploy the nginx service without Type=LoadBalancer, we will (should) stop trying to create a load balancer, and you should get out of the AWS API penalty-box.

Sorry about this, and thanks for the report.  Looks like this is pretty important to fix.

Do you happen to know why you can't create an ELB?  My guess is that you have hit your quota?

**Comment by evanphx** [comment]
@justinsb I'll remove the service, wait for the penalty-box to expire, then build it again so that we can see what the real error was.

**Comment by evanphx** [comment]
Adding the LoadBalancer service back in, I'm seeing:

```
zero :: k8/kubernetes> kubectl get events
FIRSTSEEN                         LASTSEEN                          COUNT     NAME       KIND      SUBOBJECT   REASON                         SOURCE                  MESSAGE
Wed, 15 Jul 2015 13:02:39 -0700   Wed, 15 Jul 2015 13:02:39 -0700   1         my-nginx   Service               creating loadbalancer failed   {service-controller }   failed to create external load balancer for service default/my-nginx: error tagging security group: InvalidGroup.NotFound: The security group 'sg-bd01c3d9' does not exist
                                  status code: 400, request id: []
Wed, 15 Jul 2015 13:02:45 -0700   Wed, 15 Jul 2015 13:03:35 -0700   10        my-nginx   Service             creating loadbalancer failed   {service-controller }   failed to create external load balancer for service default/my-nginx: InvalidGroup.Duplicate: The security group 'k8s-elb-a71736eae2b2c11e593680276d36fbdd' already exists for VPC 'vpc-f661fa93'
                                  status code: 400, request id: []

```

**Comment by evanphx** [comment]
I took a few different steps and seem to have gotten it both working in different ways:

First I deleted the service because it was erroring, but I noticed the count was still going up. So I deleted the security group it said was a duplicate and then it finished setting the ELB up! That meant I needed to manually clean up the ELB resources and security groups because k8 thought it was already deleted.

After cleaning up all the resources in AWS again, I added the service again and it seemed to work fine! I'm not sure how/why it initially confused itself with the duplicate SG, but it doesn't necessarily seem to happen every time.

The code could likely just, in the event that it sees the SG already exists, continue on with the setup. I could send you a PR to handle that if you'd like!

**Comment by justinsb** [comment]
I have a PR that I _think_ covers this: #11029.   I thought it was a one-off error, but it sounds like it wasn't.

Thanks so much for pausing & retrying also - that was super-helpful.  Even if we think #11029 covers this, let's leave this issue open to track the need for a backoff to avoid overloading the AWS API.  (And presumably this aggressive retrying happens on other clouds also, though maybe they have overload a sleep built in).

**Comment by evanphx** [comment]
Certainly looks like #11029 should fix it. I'm still wondering why I'd see it failing literally for hours in a loop though. If it was a problem with AWS being eventually consistent, then after an hour I'd think that first check, looking to see if the securityGroup existed, would start to return the group!

I don't know the k8 code well yet, though so perhaps it was looping in a different place that wasn't running that code to see if the security group already existed.

**Comment by justinsb** [comment]
@evanphx you're right; my explanation wasn't entirely correct...

So security group names are unique per VPC.  We also have a per-cluster name, and we filter all our AWS API interactions to only select resources tagged with this name.

If we had a security group name collision across two k8s clusters in the same VPC, that would be bad.  Thankfully the security group name we're creating includes a UUID, so this "won't happen".

However, the problem is that we created the security group, and then failed to tag it.  _But_ we were filtering based on our cluster.  So our "describe security groups" operation would _never_ see the SG, because it wasn't appropriately tagged.

We have a few workarounds to make this much less likely, and I have opened the "big picture" issue #11882.  And I'm going to think about whether we should in fact be filtering when we DescribeSecurityGroups - I think the answer is know in the case where the name has a UUID in it.

**Comment by evanphx** [comment]
@justinsb Ah! That totally makes sense. The race between resource creation and resource tagging can created "invisible" resources if the process of tagging fails. Maybe an anti-entropy function that can look at all the resources and normalize them. In the case of invisible resources, you can just delete them and hope that recreating them won't fail the tagging process.

**Comment by frosenberg** [comment]
I see this as well when frequently deploying/testing. Unfortunately, it happens far too often (running v1.0.0).

**Comment by mcadam** [comment]
+1

**Comment by jacobtomlinson** [comment]
I've been having a similar issue to this, Kubernetes can't create a load balancer as it says the security group already exists. The workaround I've found is to manually delete the security group and wait for Kubernetes to successfully create it again.

I've generally had the problem when quickly deleting a service with a `LoadBalancer` and then creating it again (I see now I should use `kubectl replace`). I wonder if it is something to do with AWS reassigning the same load balancer back to you if you haven't finished deleting it coupled with the tagging issue mentioned earlier.

**Comment by aronchick** [comment]
+1

**Comment by quinton-hoole** [comment]
cc @justinsb

**Comment by davidopp** [comment]
I had an email discussion with @justinsb on this and he said that what's remaining shouldn't be considered a blocker for 1.1, so I'll move it to 1.2-candidate.

**Comment by davidopp** [comment]
@justinsb Can you summarize what's remaining to do for this issue? Should we just close it in favor of #11882?

**Comment by davidopp** [comment]
@justinsb I see you added the v1.2 milestone -- will you be able to look into this in the next few days?
